prov/gni/src/gnix_auth_key.c

```diff
@@ -36,13 +36,87 @@
 
 #include "gnix_auth_key.h"
 #include "gnix_hashtable.h"
+#include "gnix_shmem.h"
+#include "gnix_bitmap.h"
 #include "gnix.h"
 
 #define GNIX_AUTH_KEY_HASHSEED 0xdeadbeef
 
+#define GAI_VERSION(major, minor) (((major) << 16) + (minor))
+typedef enum gnix_global_auth_info_version {
+	GNIX_GAI_VERSION_1 = GAI_VERSION(1, 0), /* initial version */
+} gnix_global_auth_info_version_t;
+
+#define MAX_VMDH_TAGS 4
+#define MAX_VMDH_REGS 4096
+#define BITMAP_ARR_SIZE(nbits) \
+	(GNIX_BITMAP_BLOCKS(nbits) * sizeof(gnix_bitmap_block_t))
+#define GNIX_VMDH_PATH "/tmp/gnix_vmdh_info"
+
+struct gnix_global_ptag_info {
+	gnix_bitmap_t prov;
+	gnix_bitmap_t user;
+	uint8_t arr[BITMAP_ARR_SIZE(MAX_VMDH_REGS)];
+};
+
+struct gnix_global_vmdh_info {
+	uint32_t version;
+	uint32_t size;
+	uint32_t table_size;
+	fastlock_t lock;
+	int ptag_index[MAX_VMDH_TAGS];
+	struct gnix_global_ptag_info info[MAX_VMDH_TAGS];
+};
+
 /* Global data storage for authorization key information */
+/* gnix_vmdh_info is the shared memory synchronization area for ptag info */
+static struct gnix_shared_memory __gnix_shmem_region;
+static struct gnix_global_vmdh_info *__gnix_vmdh_info;
 gnix_hashtable_t __gnix_auth_key_ht;
 
+static int __gnix_global_vmdh_info_init(const char *path,
+	uint32_t size, void *buffer)
+{
+	struct gnix_global_vmdh_info *info =
+		(struct gnix_global_vmdh_info *) buffer;
+	int i;
+
+	memset(info, 0, sizeof(struct gnix_global_vmdh_info));
+
+	info->version = GNIX_GAI_VERSION_1;
+	info->size = size;
+	info->table_size = _gnix_bitmap_get_buffer_size(MAX_VMDH_REGS);
+	for (i = 0; i < MAX_VMDH_TAGS; i++)
+		info->ptag_index[i] = -1;
+	fastlock_init(&info->lock);
+
+	return 0;
+}
+
+static int _gnix_open_vmdh_info_file(const char *path)
+{
+	int ret;
+
+	if (!__gnix_vmdh_info) {
+		ret = _gnix_shmem_create(path,
```

Comment:
If the file does exist, it is opened with exclusive access. The opener will create the memory structure and exist. Other files who attempt to open the file while in exclusive access by the owner, will wait until the file is the expected size, then attempt to open it after that condition is satisfied. 

Inside the memory segment, there is a lock for serializing access to critical sections. 

See gnix_shmem.c:60 , gnix_shmem.c:138 , gnix_auth_key.c:91, gnix_auth_key.c:168